File: client_utils/src/net_sim.rs
//! A deterministic network simulator for testing and demonstrating netcode.
//!
//! Prediction, reconciliation, and interpolation are only interesting under
//! latency, so exercising them means injecting delay, jitter, and loss in a
//! *reproducible* way. [`LatencyLink`] is a one-way time-ordered delay queue,
//! and [`Rng`] is a tiny seeded PRNG so a run repeats exactly.
//!
//! Gated behind the `net-sim` feature; it is a test and demo aid, not part of
//! the client API. The `netcode_playground` example uses it as its network.

use std::collections::VecDeque;

/// A tiny deterministic PRNG (xorshift64). Not for anything but reproducible
/// jitter and loss.
#[derive(Clone, Debug)]
pub struct Rng(u64);

impl Rng {
  /// Seeds the generator. A fixed seed makes a run repeatable.
  pub fn new(seed: u64) -> Self {
    // Avoid the zero state, which xorshift cannot escape.
    Self(seed | 1)
  }

  fn next_u64(&mut self) -> u64 {
    let mut x = self.0;
    x ^= x << 13;
    x ^= x >> 7;
    x ^= x << 17;
    self.0 = x;
    x
  }

  /// A float in `[0, 1)`.
  pub fn unit(&mut self) -> f32 {
    (self.next_u64() >> 40) as f32 / (1u64 << 24) as f32
  }

  /// An integer in `[0, n]` inclusive.
  pub fn up_to(&mut self, n: u64) -> u64 {
    if n == 0 {
      0
    } else {
      self.next_u64() % (n + 1)
    }
  }
}

/// One direction of a simulated wire: packets handed in at `now` become
/// deliverable at `now + latency (+ jitter)`, unless dropped. Generic over the
/// packet type, drive both directions with two of them.
#[derive(Debug)]
pub struct LatencyLink<T> {
  queue: VecDeque<(u64, T)>,
}

impl<T> LatencyLink<T> {
  pub fn new() -> Self {
    Self { queue: VecDeque::new() }
  }

  /// Hands a packet to the wire. It may be delayed by `latency_ms` plus up to
  /// `jitter_ms`, or dropped with probability `loss_pct / 100`.
  ///
  /// Jitter can reorder delivery; [`drain_due`](Self::drain_due) sorts by
  /// delivery time so the receiver still gets them in order.
  pub fn send(&mut self, now_ms: u64, packet: T, latency_ms: u64, jitter_ms: u64, loss_pct: f32, rng: &mut Rng) {
    if loss_pct > 0.0 && rng.unit() * 100.0 < loss_pct {
      return;
    }
    let deliver_at = now_ms + latency_ms + rng.up_to(jitter_ms);
    self.queue.push_back((deliver_at, packet));
  }

  /// Removes and returns every packet whose delivery time has arrived, oldest
  /// delivery first.
  pub fn drain_due(&mut self, now_ms: u64) -> Vec<T> {
    let mut due: Vec<(u64, T)> = Vec::new();
    let mut kept: VecDeque<(u64, T)> = VecDeque::new();
    for (at, packet) in self.queue.drain(..) {
      if at <= now_ms {
        due.push((at, packet));
      } else {
        kept.push_back((at, packet));
      }
    }
    self.queue = kept;
    due.sort_by_key(|(at, _)| *at);
    due.into_iter().map(|(_, p)| p).collect()
  }

  pub fn in_flight(&self) -> usize {
    self.queue.len()
  }
}

impl<T> Default for LatencyLink<T> {
  fn default() -> Self {
    Self::new()
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn a_packet_is_held_until_its_latency_elapses() {
    let mut link = LatencyLink::new();
    let mut rng = Rng::new(1);
    link.send(0, "hi", 100, 0, 0.0, &mut rng);

    assert!(link.drain_due(50).is_empty(), "not yet due");
    assert_eq!(link.drain_due(100), vec!["hi"], "due at latency");
  }

  #[test]
  fn full_loss_drops_everything() {
    let mut link = LatencyLink::new();
    let mut rng = Rng::new(2);
    for _ in 0..100 {
      link.send(0, 1u8, 10, 0, 100.0, &mut rng);
    }
    assert_eq!(link.in_flight(), 0, "every packet dropped at 100% loss");
  }

  #[test]
  fn reordered_deliveries_come_out_in_time_order() {
    let mut link = LatencyLink::new();
    let mut rng = Rng::new(3);
    // Deliberately enqueue a later-delivered packet first.
    link.queue.push_back((200, "b"));
    link.queue.push_back((100, "a"));
    assert_eq!(link.drain_due(300), vec!["a", "b"]);
    let _ = &mut rng;
  }

  #[test]
  fn the_rng_is_reproducible() {
    let mut a = Rng::new(42);
    let mut b = Rng::new(42);
    assert_eq!(a.up_to(1000), b.up_to(1000));
    assert_eq!(a.unit(), b.unit());
  }

  #[test]
  fn under_heavy_jitter_every_packet_arrives_exactly_once() {
    use std::collections::HashSet;
    let mut link = LatencyLink::new();
    let mut rng = Rng::new(0xABCD);

    // 500 packets, each tagged with its unique send time, heavy jitter, no loss.
    let mut sent = 0;
    for now in (0..5000).step_by(10) {
      link.send(now, now, 50, 200, 0.0, &mut rng);
      sent += 1;
    }

    // Drain past the last possible delivery.
    let mut seen: HashSet<u64> = HashSet::new();
    for p in link.drain_due(1_000_000) {
      assert!(seen.insert(p), "packet {p} delivered twice");
    }
    assert_eq!(seen.len(), sent, "every packet arrived, none lost despite reordering");
    assert_eq!(link.in_flight(), 0, "nothing left stuck in the queue");
  }
}

File: flutter/plaza_client_utils/lib/src/net_sim.dart
import 'dart:collection';

/// A tiny deterministic PRNG (xorshift64). Not for anything but reproducible
/// jitter and loss.
///
/// `>>>` rather than `>>`, and the same reason as `mix64`: Dart's `>>`
/// sign-extends and this is a u64 algorithm. The sequence agrees with the Rust
/// `Rng` for the same seed, which is what makes a Rust scenario and a Dart one
/// comparable at all.
///
/// Ported from `plaza_client_utils::net_sim::Rng`.
class Rng {
  /// A fixed seed makes a run repeatable.
  Rng(int seed) : _state = seed | 1;

  int _state;

  int _nextU64() {
    var x = _state;
    x ^= x << 13;
    x ^= x >>> 7;
    x ^= x << 17;
    _state = x;
    return x;
  }

  /// A double in `[0, 1)`.
  ///
  /// Drawn from the top 24 bits, so the quotient is exact in both `f32` and
  /// `double` and agrees with Rust exactly rather than approximately.
  double unit() => (_nextU64() >>> 40) / 16777216.0;

  /// An integer in `[0, n]` inclusive.
  int upTo(int n) {
    if (n <= 0) return 0;
    return _unsignedMod(_nextU64(), n + 1);
  }

  /// `x` as a u64, modulo `m`.
  ///
  /// Dart's `%` on a negative left operand is the mathematical modulo of a
  /// *signed* value, and half of every PRNG's draws have the top bit set. So the
  /// negative case is folded back by hand: `u = 2^63 + r`, and `2^63 mod m` is
  /// accumulated by doubling, never exceeding `m`.
  ///
  /// Requires `m <= 2^62`, which every jitter value is by a wide margin.
  static int _unsignedMod(int x, int m) {
    if (x >= 0) return x % m;
    final r = x & 0x7FFFFFFFFFFFFFFF;
    var p = 1;
    for (var i = 0; i < 63; i++) {
      p = (p * 2) % m;
    }
    return (p + r % m) % m;
  }
}

/// Whether the simulated wire may deliver packets out of send order.
///
/// This is not a detail. Impairment tooling that can produce failures the real
/// transport cannot is worse than no tooling, because the failures are credible
/// enough to spend a day chasing, and because it quietly hides that the real
/// system has stronger guarantees than the tests assume. Pick the one that matches
/// the transport being stood in for.
enum PacketOrdering {
  /// Jitter delays a packet, possibly past its successors, but never ahead of its
  /// predecessors. **The default**, and what TCP, WebSocket, QUIC streams and any
  /// ordered channel actually do.
  ordered,

  /// Jitter may reorder freely, so a later packet can arrive first. What raw UDP
  /// and unordered datagram channels do. Choose this deliberately: a delta stream
  /// that assumes ordering will diverge under it, which is a real finding on a
  /// datagram transport and a phantom on an ordered one.
  unordered,
}

/// One direction of a simulated wire: packets handed in at `now` become
/// deliverable at `now + latency (+ jitter)`, unless dropped. Generic over the
/// packet type, so drive both directions with two of them.
///
/// Ported from `plaza_client_utils::net_sim::LatencyLink`.
class LatencyLink<T> {
  LatencyLink({this.ordering = PacketOrdering.ordered});

  /// Whether jitter may reorder. See [PacketOrdering].
  final PacketOrdering ordering;

  final Queue<(int, T)> _queue = Queue<(int, T)>();

  /// The delivery time of the last packet queued, so an ordered link can hold a
  /// jittered packet back behind its predecessor.
  int _lastDeliverMs = 0;

  /// Hands a packet to the wire. It may be delayed by [latencyMs] plus up to
  /// [jitterMs], or dropped with probability `lossPct / 100`.
  ///
  /// Under [PacketOrdering.ordered] the jittered delivery time is clamped to at
  /// least the previous packet's, so jitter shows up as a packet arriving late
  /// rather than as the stream shuffling. Loss is independent of ordering: an
  /// ordered transport still loses whole connections and, at this level of
  /// abstraction, still models a dropped application message.
  void send(
    int nowMs,
    T packet, {
    required int latencyMs,
    int jitterMs = 0,
    double lossPct = 0.0,
    required Rng rng,
  }) {
    if (lossPct > 0.0 && rng.unit() * 100.0 < lossPct) return;
    var deliverAt = nowMs + latencyMs + rng.upTo(jitterMs);
    if (ordering == PacketOrdering.ordered && deliverAt < _lastDeliverMs) {
      deliverAt = _lastDeliverMs;
    }
    _lastDeliverMs = deliverAt;
    _queue.addLast((deliverAt, packet));
  }

  /// Removes and returns every packet whose delivery time has arrived, oldest
  /// delivery first.
  List<T> drainDue(int nowMs) {
    final due = <(int, T)>[];
    final kept = Queue<(int, T)>();
    for (final entry in _queue) {
      if (entry.$1 <= nowMs) {
        due.add(entry);
      } else {
        kept.addLast(entry);
      }
    }
    _queue
      ..clear()
      ..addAll(kept);
    // Stable, so packets sharing a delivery time keep their send order, which is
    // what an ordered link promises after clamping.
    mergeSortByDeliveryTime(due);
    return due.map((e) => e.$2).toList(growable: false);
  }

  /// Enqueues a packet at an exact delivery time, bypassing latency, jitter and
  /// loss. For tests that need a specific arrival order.
  void enqueueAt(int deliverAtMs, T packet) {
    _lastDeliverMs = deliverAtMs > _lastDeliverMs ? deliverAtMs : _lastDeliverMs;
    _queue.addLast((deliverAtMs, packet));
  }

  int get inFlight => _queue.length;
}

/// A stable sort by delivery time.
///
/// Dart's `List.sort` is not stable, and an unstable sort here would reorder
/// packets that share a delivery time, which is exactly the shuffling
/// [PacketOrdering.ordered] exists to rule out.
void mergeSortByDeliveryTime<T>(List<(int, T)> items) {
  if (items.length < 2) return;
  final buffer = List<(int, T)>.of(items);
  void merge(int lo, int mid, int hi) {
    var i = lo;
    var j = mid;
    for (var k = lo; k < hi; k++) {
      if (i < mid && (j >= hi || buffer[i].$1 <= buffer[j].$1)) {
        items[k] = buffer[i++];
      } else {
        items[k] = buffer[j++];
      }
    }
  }

  void sort(int lo, int hi) {
    if (hi - lo < 2) return;
    final mid = lo + (hi - lo) ~/ 2;
    sort(lo, mid);
    sort(mid, hi);
    buffer.setRange(lo, hi, items, lo);
    merge(lo, mid, hi);
  }

  sort(0, items.length);
}
